### DISCUSSION TITLE
Search does not work on postgres citext fields

### DISCUSSION THREAD

**Discussion by bamboowonder** [directus/directus#13758] [Feature Requests]
### Describe the Bug

I've finally diagnosed why search is not working for us.
varchar fields are fine.
text fields are fine.
citext fields are not searched at all.  app or api.

Previously a bug report had fixed it so that Directus treated "citext" as a "text" field instead of "unknown".

But it looks like the search mechanism doesn't include database citext fields.  Even though Directus sees them as text.

### To Reproduce

create a varchar, a text, a citext field in database with test data.
search on the values from any of those 3 fields using the search box in directus or the api
nothing in citext is searched

### Errors Shown

no errors are shown anywhere

### What version of Directus are you using?

9.12.1

### What version of Node.js are you using?

12.22.9

### What database are you using?

postgres 14

### What browser are you using?

chrome

### How are you deploying Directus?

CLI npm package

**Comment by bamboowonder** [comment]
To followup.   I complete repaved the Directus instance.   And found a sortof related bug, in that I modified a citext field using Directus, and after commiting the changes (to make it NOT NULL), I noticed that all of a sudden the "search" worked for that field.    However I noticed that Directus when it altered the table to add not null, it also changed the field type to text from citext.   (Which I guess makes sense if allowing Directus to manage the database).

What was interesting, is that even with the column being text, case-INsensitive search functioned!   Whereas a few versions ago it didn't function.

So I'm not sure if I should close this bug.  But I sure appreciate that Directus now supports case-insensitive search even on postgres without having to specify citext fields!

**Comment by azrikahar** [comment]
This is an interesting one. Seems like the reason why it wasn't working initially is due to the search logic checks the field type to actually add said field into the WHERE query:

https://github.com/directus/directus/blob/3e1c753573315c2bd9bb2d4ac38223a51d3ce0fe/api/src/utils/apply-query.ts#L613-L614

however, during this time (even though the App shows `text` already), the field type of the citext field in this IF statement is actually `unknown`, so it was never added to the search query.

Tracing it back to the `data_type` of the column over here:

https://github.com/directus/directus/blob/3e1c753573315c2bd9bb2d4ac38223a51d3ce0fe/packages/schema/src/dialects/postgres.ts#L32-L35

will retrieve `USER-DEFINED` instead. This should be related to this comment here - https://github.com/knex/knex-schema-inspector/pull/73#issuecomment-946869695, pointing out the quote from postgres:

> Data type of the array elements, if it is a built-in type, else USER-DEFINED (in that case, the type is identified in udt_name and associated columns).

but as @bamboowonder described, "updating the field" via disabling then re-enabling `Allow NULL value` makes it properly returning `text` and not `USER-DEFINED` anymore going forward. 

Feels like this is not really a Directus bug though, and could something to be explored in the knex-schema-inspector end, additionally since citext is Postgres-specific and there already is case-insensitive functionality built-in now. @rijkvanzanten thoughts on this?

---

- Before "updating" the citext field by toggle off and on the nullable property, we'll get `USER-DEFINED` for `data_type`:

   ![chrome_80T0WBwUGP](https://user-images.githubusercontent.com/42867097/172117791-3c9a21a7-0e9e-4842-875b-5d59fc71d752.png)

- After "updating" the field:

   ![chrome_f1k5VsonYM](https://user-images.githubusercontent.com/42867097/172117813-7915b7fb-d62c-41a9-b7e4-538eca865ca1.png)

**Comment by rijkvanzanten** [comment]
@azrikahar You were the one adding support for search in citext columns in https://github.com/directus/directus/pull/12670 🥳 

Closing this, as I believe the original issue described is working as intended. When it comes to adding native support for citext (which is a postgres extension, rather than a core feature) I think we should keep discussing that in a new discussion

**Comment by bamboowonder** [comment]
> Closing this, as I believe the original issue described is working as intended.

The thing is though, it isn't working as intended.   Directus search does *not* search any citext fields (even though Directus correctly recognizes citext as text in the interface.

**Comment by rijkvanzanten** [comment]
Ah I see, I got confused by the 

> What was interesting, is that even with the column being text, case-INsensitive search functioned! Whereas a few versions ago it didn't function.

👍🏻

**Comment by rijkvanzanten** [comment]
`citext` is a custom user-defined type injected through an extension, so I think we should pick this up as a feature request to implement support for PG custom types and how to handle the type abstraction for those, rather than trying to hack in support for potentially common ones at a time 🤔 Moving this to a discussion 👍🏻

**Comment by bamboowonder** [comment]
Fair enough.  I had only really used citext because Directus didn't support case-insensitive search.

**But now it only "sortof" works.   So we're left in limbo.

As an example, if we leave a db column as citext (Directus not managing that field), then Search doesn't work at all for that field.  but case-insensitive Filter *does* work.

If I let Directus manage that field, it arbitrarily changes the type *in the database* from citext to text.    Now the built-in case-insensitive Search works.   However,  Filtering is now only case-sensitive.

A little like playing wack-a-mole :)

**Reply by bamboowonder** [reply]
I'd love to just let Directus/Knex be opinionated, and match our field types to the functionality that Directus offers.   (meaning ditch the citext, and make use of either text or varchar in places where I would otherwise use text for everything)

The only way I can do that is if Directus Search and Filter both support case-insensitivity out-of-the-box.

I imagine for anyone using Postgres, this wish falls under the 80/20 criteria?

Thanks a bunch for a project I otherwise love working with.

**Reply by rijkvanzanten** [reply]
> The only way I can do that is if Directus Search and Filter both support case-insensitivity out-of-the-box.

Search has been case insensitive from the start (it uses a `lower(column) LIKE %searchquery%` type search), dedicated case insensitive filter operators were actually just added in https://github.com/directus/directus/pull/11737 🙂

**Reply by bamboowonder** [reply]
It looks like case insensitive Filter was added as an option to string/varchar.   
It wasn't added to text though. 
Is there any possibility of that happening (in postgres text is identical to varchar without a limit)

**Reply by rijkvanzanten** [reply]
I don't think the API currently differs the filter operators between strings and text does it? Or are you referring to the operators that show in the app (in which case that's https://github.com/directus/directus/pull/13431 )

**Reply by bamboowonder** [reply]
Yes I was referring to the app, and it looks like that #13431 is in the oven.  Cheers thank you

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
